src/helpers/checkbox.ts

```diff
@@ -78,10 +78,18 @@ export function getNextCheckboxState({
     return nextCheckboxState;
   }
 
-  // Some, or all selected.
+
+  // if `all-selected`, just clear the set
+  if (checkboxState[checkboxProp].state === 'all-selected') {
```

Comment:
Hmmm, apparently this breaks the test in uncontrolled table: if at the moment all data are selected, then if we unselect one checkbox, then all checkbox will be unselected instead (see the [test log in the job](https://github.com/imballinst/react-bs-datatable/actions/runs/4535917307/jobs/7993782140?pr=174)).

From my understanding, _usually_ uncontrolled tables are purely uncontrolled, while at the moment the case that we want to solve here is, the table is actually using a controlled pagination from the backend. I thought of this over last weekend, the thoughtflow is like this:

> If in controlled table we are able to clear selection of all data (that means, clearing the `selected` field), then we should also be able to add all data to selection

But adding the entire IDs isn't possible when we're using a backend for pagination, because the table doesn't know the data IDs of the rest of the table pages. The "context" that the controlled table is able to see is only the current page only.

Perhaps, the solution here is by providing a "hook" to manually set the state here?

https://github.com/imballinst/react-bs-datatable/blob/af2f336a2e34a8746a71f6c60377bf5b545a2009/src/helpers/hooks.ts#L133-L146

Something like: 

```ts
function createBulkCheckboxClickHandler(
  type?: GetNextCheckboxStateParams['type'],
  checkboxInfo?: { idProp: string; checkboxProp: string },
  newStateOverrider?: (newState: Record<string, CheckboxState>) => Record<string, CheckboxState>
) {
  // ...
  let nextCheckboxState = getNextCheckboxState({ /*...*/ })
  if (newStateOverrider) {
    nextCheckboxState = newStateOverrider(nextCheckboxState)
  }
}
```

I think it's a perfect case if, so we can do this:

```ts
createBulkCheckboxClickHandler("remove", {
  idProp: "id",
  checkboxProp: "checkbox"
}, (newCheckboxState) => {
  return { ...newCheckboxState, checkbox: { selected: new Set(), state: 'none-selected' } }
})
```

Let me know what you think about it.

Reply:
That makes sense to me. This feels like the config overriding pattern that a bundler like Webpack would use. If you can do something like this, that would be great 👍